core/src/main/scala/kafka/tools/GetOffsetShell.scala

```diff
@@ -222,11 +240,19 @@ object GetOffsetShell {
   }
 
   /**
-   * Return the partition infos. Filter them with topicPartitionFilter.
+   * Return the partition infos. Filter them with topicFilter and topicPartitionFilter.
    */
-  private def listPartitionInfos(consumer: KafkaConsumer[_, _], topicPartitionFilter: PartitionInfo => Boolean): Seq[PartitionInfo] = {
-    consumer.listTopics.asScala.values.flatMap { partitions =>
-      partitions.asScala.filter(topicPartitionFilter)
+  private def listPartitionInfos(client: Admin,
+                                 topicFilter: String => Boolean,
+                                 topicPartitionFilter: PartitionInfo => Boolean,
+                                 excludeInternalTopics: Boolean): Seq[PartitionInfo] = {
+    val topics = client.listTopics(new ListTopicsOptions().listInternal(!excludeInternalTopics)).names().get().asScala.filter(topicFilter)
+    client.describeTopics(topics.asJavaCollection).allTopicNames().get().asScala.flatMap { case (topic, description) =>
+      description
+        .partitions()
+        .asScala
+        .map(tp => new PartitionInfo(topic, tp.partition(), tp.leader(), tp.replicas().asScala.toArray, tp.isr().asScala.toArray))
```

Comment:
Instead of converting to `PartitionInfo`, could we adapt the filters to work with what we have here?

Reply:
I tried this according to your suggestion, however, the class`TopicPartitionInfo` holds no topic information and the class `TopicPartition` holds no leader information, these 2 fields are necessary for `listOffsets`, so I just left as it was.